### PR TITLE
tests, dual stack: split probes tests to test per IP-Family

### PR DESCRIPTION
**Pull request by alonSadan**
**What this PR does / why we need it**:
some of the dual stack tests test both IPv4 and IPv6 and iterate over
the IP addresses and act accordingly. A better way will be to
test each protocol separately.

This PR splits tests which tests both protocols, to separate tests, or
table entries.

Signed-off-by: alonSadan <asadan@redhat.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:95]: You should add this on lines 62-78 instead, and remove this part (so we keep the positive tests before the negative ones).

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:170]: same here; you should do this on https://github.com/kubevirt/kubevirt/pull/4341/files#diff-f2964358d14f79c439288df507310df44dbc0de205c1ec0a02a78f9d4cbe7c13R137 instead.

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:258]: I'm not sure you should use this on IPv4 - e.g. you could return the original probe on IPv4.

Nothing wrong with being specific here; always setting the IP of the host is probably a good thing (but I think it is important to also test the behavior when we omit the hostIP on the probe definition, so we test the 'default' argument).

**Review by maiqueb — CHANGES_REQUESTED** [review]

**Comment by alonSadan** [comment]
If this PR won't land, then #4420 should be instead

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:232]: >  always setting the IP of the host is probably a good thing (but I think it is important to also test the behavior when we omit the hostIP on the probe definition, so we test the 'default' argument).

so I left it as it is, and added another case where no ip family is specified. in that case we won't configure the probe with host IP.
What do you think?

**Review by alonSadan — COMMENTED** [review]
Thank you @maiqueb for the review, and I am very sorry for the late replay.
I added another case where no ip family is specified and in that case the probe won't be configured with host ip. 

Also, now we just use the backend pod as the server and not use a server on the vmi, but I am not sure this is what you meant.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go]: Done. Removed the whole create-server-on-vmi part.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go]: Done. Removed the whole create-server-on-vmi part.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:232]: @maiqueb  the default (no hostIp) [doesn't seem](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4341/pull-kubevirt-e2e-k8s-1.18/1321443971087470592#1:build-log.txt%3A5281) to work.

doing this locally gives:
``` 
Warning  Unhealthy  85s (x13 over 2m25s)  kubelet, node01    Readiness probe failed: dial tcp 10.244.196.158:1500: connect: connection refused
```
event in the virt-launcher pod. Can you see what I did wrong?

Comment (maiqueb) [reply tests/probes_test.go:232]: When you don't set the ipFamily in the service, it defaults to the pod IP address.

Since the server is now located in a sidekick pod, the probes will not work (the server is in the sidekick pod, and kubelet is attempting to contact the VM via the virt-launcher IP address).

You should start the server on the VMI on an empty IPFamily (not creating the sidekick pod).

**Review by maiqueb — CHANGES_REQUESTED** [review]

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:232]: Understand. will do.

What confused me is that I couldn't accept the fact that the vmi is not being checked. now that I know that it is on purpose, it is a lot more clear. Thanks for explaining.

**Comment by alonSadan** [comment]
fails on https://github.com/kubevirt/kubevirt/issues/4385

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:246]: Done.

File: tests/probes_test.go
Comment (RamLavi) [inline tests/probes_test.go:166]: since ipv4 case works with `tests.StartHTTPServer` I suggest we use it in this case.
```
table.Entry("[test_id:1199][posneg:positive]with working TCP probe and tcp server on ipv4", tcpProbe, v1.IPv4Protocol, tests.StartHTTPServer),
```

File: tests/probes_test.go
Comment (RamLavi) [inline tests/probes_test.go:137]: since ipv4 case works with tests.StartHTTPServer I suggest we only use the pod-sidekick in ipv6 case:
```
if IPFamily == v1.IPv6Protocol {
```

File: tests/probes_test.go
Comment (RamLavi) [inline tests/probes_test.go:143]: can you add an explanation here? something like
```
By("Attaching the readiness probe to an external pod server")
```

**Review by RamLavi — CHANGES_REQUESTED** [review]
added some comments to increase readability.

Also, I'm not sure where exactly we are checking that the vmi is "ready"
I see that `createReadyCirrosVMIWithReadinessProbe` is blocking until the vmi is `started` but does that mean that the vmi is "ready"?

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go]: Done.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:140]: Done.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go]: Done.

**Review by alonSadan — COMMENTED** [review]
Thanks Ram for the review

**Comment by RamLavi** [comment]
thanks
/lgtm

**Review by maiqueb — CHANGES_REQUESTED** [review]

File: tests/probes_test.go
Comment (dhiller) [inline tests/probes_test.go:98]: @alonSadan I think what you want this table entry to be called is this.
```suggestion
			table.Entry("[test_id:1202][posneg:positive]with working HTTP probe and tcp server,no ip family specified ", httpProbe, *new(v1.IPFamily)),
```

As the merge tool relies on unique naming during processing it thinks that a test has run more than once.

Comment (alonSadan) [reply tests/probes_test.go]: Thank you! 
what a stupid mistake.
Fixed.

**Comment by alonSadan** [comment]
[removed](https://github.com/kubevirt/kubevirt/compare/2d99f3ee7abf7f7f42f74fa8d36813d6d21c74ed..2416cd143f17429fa843e612bf31c2d73965b23e) the call to the server starter from the test function signature, and removed the go.mod leftover .

[fixed](https://github.com/kubevirt/kubevirt/compare/7473f50cff7a0d24079f31f29c5582154f7b9240..6ef6aae33cdfa359693093004dada7e54658836f) the table entries naming. to avoid https://github.com/kubevirt/kubevirt/issues/4385

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:76]: You can encapsulate these in an helper function. Hopefully that will improve readability and remove the need for the `By` clause.

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:95]: Could you export `*new(v1.IPFamily)` as a const ?

**Review by maiqueb — COMMENTED** [review]
Nice. Really close.

You should start looking into finding an approver.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:76]: Done.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:95]: exported it as a a var since `*new(V1.IPFamily)` is not a constant.

**Comment by alonSadan** [comment]
Thanks @maiqueb. [I made the changes](https://github.com/kubevirt/kubevirt/compare/4cd29fbaf3695301cea56271953436147cb2ac62..f6cda312d502e5e7e15a941cadbe93ee994de6fd) you asked.
I accidentally  mixed it with the rebase changes. I hope it is OK.
I exported  attaching the pod with the probe to a helper, and used a var for the non determined ipvFamily.

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:287]: why liveness probe ? this should be used also for readiness probes.

File: tests/probes_test.go
Comment (maiqueb) [inline tests/probes_test.go:282]: This name is not OK; `pointProbeToSupportPod` is, imho, better.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Some naming nits.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:25]: But it is not accurate to call it "default" -  https://play.golang.org/p/v31J0aOqFhl  .
It just implies that we will use the default. so calling it default could be a bit misleading. What do you think?

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:287]: Damm. Forgot to change that. Thanks for noticing.
 Done.

**Comment by alonSadan** [comment]
[changed](https://github.com/kubevirt/kubevirt/compare/f6cda312d502e5e7e15a941cadbe93ee994de6fd..5620c5e2690e5ec3b61ce4ad7befe03b49a57911) thenames as @maiqueb  suggested, except for the blankIPFamily which is still in question.

please ignore the go.mod changes they are dropped in the [next push](https://github.com/kubevirt/kubevirt/compare/5620c5e2690e5ec3b61ce4ad7befe03b49a57911..3431cada38a28d86b7013d7d68714c5532babdf7)

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go]: Much better. Done.

File: tests/probes_test.go
Comment (maiqueb) [reply tests/probes_test.go:25]: If you can only get the attention of a maintainer after an lgtm, I'll give it to you.

Comment (alonSadan) [reply tests/probes_test.go:25]: haha
Thanks.
@AlonaKaplan said she might will, 
but I'll need the lgtm first...

**Comment by alonSadan** [comment]
> /lgtm
> 
> (I'm assuming the tests are passing).

Thank you:)

**Comment by alonSadan** [comment]
just [rebased](https://github.com/kubevirt/kubevirt/compare/3431cada38a28d86b7013d7d68714c5532babdf7..31063b23cb7e840b4bb8342576efff6417c6610c)

File: tests/probes_test.go
Comment (phoracek) [inline tests/probes_test.go:147]: This talks about readiness probe while the test is mean for liveness.

File: tests/probes_test.go
Comment (phoracek) [inline tests/probes_test.go:150]: Ditto

**Review by phoracek — CHANGES_REQUESTED** [review]

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:147]: ooops. Done.

File: tests/probes_test.go
Comment (alonSadan) [reply tests/probes_test.go:150]: Done.

**Comment by alonSadan** [comment]
[fixed](https://github.com/kubevirt/kubevirt/compare/31063b23cb7e840b4bb8342576efff6417c6610c..21146f8096a785e6b0da133448d6ce2e7575973a) @phoracek   comments

**Review by phoracek — APPROVED** [review]
/approve
/lgtm

Plus taking over the previous lgtm, since there was a minimal change.

Thanks.

**Comment by alonSadan** [comment]
> /approve
> /lgtm
> 
> Plus taking over the previous lgtm, since there was a minimal change.
> 
> Thanks.

Thank you!

**Pull request merged**
